### PR TITLE
[Raft] Implement AppendEntries.

### PR DESCRIPTION
**Pull request by liana-yang**
@flyingmeat I'll change the base branch to the master branch of your repo after the previous pull requests are merged.

### CODE REVIEW

**Comment by flyingmeat** [comment]
Let's review this together tomorrow

File: lab/src/raft/append_entries.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	}
+
+	// If leaderCommit > commitIndex, set commitIndex = min(leaderCommit, index of last new entry)
+	if (args.LeaderCommit > rf.commitIndex) {
+		lastNewEntryIndex := -1
+		lastNewEntry := getLastLog(args.Log)
+		if lastNewEntry != nil {
+			lastNewEntryIndex = lastNewEntry.Index
+		}
+		rf.commitIndex = math.Min(args.LeaderCommit, lastNewEntryIndex)
```
Comment (flyingmeat) [inline lab/src/raft/append_entries.go]: TODO - apply Msg between lastApplied and commitIndex

File: lab/src/raft/append_entries.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	hasPrevLog := len(rf.log) >= args.PrevLogIndex + 1
+	if !hasPrevLog || rf.log[args.PrevLogIndex].Term != args.PrevLogTerm {
+		reply.Success = false
+		return
+	}
+
+	// If an existing entry conflicts with a new one (same index but different terms),
+	// delete the existing entry and all that follow it (§5.3)
+	for i := range rf.log {
+		if args.Log[i].Term != rf.log[i].Term {
```
Comment (flyingmeat) [inline lab/src/raft/append_entries.go]: log => Entries
Should also loop through leader's log

File: lab/src/raft/append_entries.go (hunk 1 above)
Comment (liana-yang) [reply lab/src/raft/append_entries.go]: Done.

File: lab/src/raft/append_entries.go (hunk 2 above)
Comment (liana-yang) [reply lab/src/raft/append_entries.go]: Done.

Comment (flyingmeat) [reply lab/src/raft/append_entries.go]: Forget push to remote?

**Review by flyingmeat — APPROVED** [review]

**Review by flyingmeat — APPROVED** [review]

**Pull request merged**